In [4]:
import os
import pandas as pd

# Set the folder path where your CSVs are stored
folder_path = 'C:/Users/HP/OneDrive/Desktop/Analysis & Visualization'

# List all CSV files in the folder
csv_files = [f for f in os.listdir(folder_path) if f.endswith('.csv')]

# Read each CSV into a dictionary of DataFrames
tables = {}
for file in csv_files:
    table_name = file.replace('.csv', '')
    tables[table_name] = pd.read_csv(os.path.join(folder_path, file))

# Print summary of loaded tables
for name, df in tables.items():
    print(f"{name}: shape={df.shape}, columns={list(df.columns)}")

GTS_CITIES: shape=(1000, 6), columns=['GTS_CITY_ID', 'GTS_CITY_NAME', 'GTS_CITY_DESCRIPTION', 'GTS_CITY_STATUS', 'GTS_COUNTRY_ID', 'GTS_STATE_ID']
GTS_COMPANIES: shape=(1000, 8), columns=['GTS_COMPANY_ID', 'GTS_COMPANY_NAME', 'GTS_COMPANY_DESCRIPTION', 'GTS_COMPANY_STATUS', 'GTS_COMPANY_CONTACT_EMAIL', 'GTS_COUNTRY_CODE', 'GTS_COMPANY_CONTACT_MOBILE_NUMBER', 'GTS_COMPANY_CONTACT_ALTERNATE_MOBILE_NUMBER']
GTS_COUNTRIES: shape=(1000, 5), columns=['GTS_COUNTRY_ID', 'GTS_COUNTRY_NAME', 'GTS_COUNTRY_DESCRIPTION', 'GTS_COUNTRY_STATUS', 'GTS_COUNTRY_CODE']
GTS_CURRENCIES: shape=(1000, 5), columns=['GTS_CURRENCY_CODE', 'GTS_CURRENCY_ID', 'GTS_CURRENCY_NAME', 'GTS_CURRENCY_DESCRIPTION', 'GTS_CURRENCY_STATUS']
GTS_DEGREES: shape=(1000, 4), columns=['GTS_DEGREE_ID', 'GTS_DEGREE_NAME', 'GTS_DEGREE_DESCRIPTION', 'GTS_DEGREE_STATUS']
GTS_EMPLOYER_COMPANY_DETAILS: shape=(1000, 6), columns=['GTS_EMPLOYER_COMPANY_DETAILS_ID', 'GTS_EMPLOYER_ID', 'GTS_EMPLOYER_COMPANY_ID', 'GTS_JOB_TITLE_ID', 'GTS_EMPLOY

Inference
- Overall, the data is well-structured and suitable for analysis.
- No critical missing values in key identifiers or outcome columns.
- Some enrichment tables (currencies, addresses, work experience) may need cleaning or imputation for advanced analysis, but this will not block the main training effectiveness analysis.

In [5]:
import os
import pandas as pd

# 1. Load all tables
folder_path = 'C:/Users/HP/OneDrive/Desktop/Analysis & Visualization'
csv_files = [f for f in os.listdir(folder_path) if f.endswith('.csv')]
tables = {}
for file in csv_files:
    table_name = file.replace('.csv', '')
    tables[table_name] = pd.read_csv(os.path.join(folder_path, file))

# 2. Deduplicate GTS_CURRENCIES
if 'GTS_CURRENCIES' in tables:
    before = tables['GTS_CURRENCIES'].shape[0]
    tables['GTS_CURRENCIES'] = tables['GTS_CURRENCIES'].drop_duplicates()
    after = tables['GTS_CURRENCIES'].shape[0]
else:
    before = after = 0

# 3. Show deduplication result and confirm key tables are loaded
key_tables = ['GTS_USERS', 'GTS_TRAININGS', 'GTS_USER_TRAININGS', 'GTS_USER_SKILLS', 'GTS_JOB_APPLICATIONS', 'GTS_EMPLOYMENT_AGREEMENTS']
key_loaded = {k: (k in tables and tables[k].shape) for k in key_tables}

{'dedup_currencies': {'before': before, 'after': after}, 'key_loaded': key_loaded}

{'dedup_currencies': {'before': 1000, 'after': 20},
 'key_loaded': {'GTS_USERS': (1000, 10),
  'GTS_TRAININGS': (1000, 7),
  'GTS_USER_TRAININGS': (1000, 10),
  'GTS_USER_SKILLS': (1000, 8),
  'GTS_JOB_APPLICATIONS': (1000, 15),
  'GTS_EMPLOYMENT_AGREEMENTS': (1000, 29)}}

Deduplication complete: GTS_CURRENCIES reduced from 1000 to 20 unique rows. All key tables for the main analysis are loaded and ready.

In [6]:
# Extract key tables for integration
users = tables['GTS_USERS']
trainings = tables['GTS_TRAININGS']
user_trainings = tables['GTS_USER_TRAININGS']
user_skills = tables['GTS_USER_SKILLS']
job_apps = tables['GTS_JOB_APPLICATIONS']
employment = tables['GTS_EMPLOYMENT_AGREEMENTS']

# Standardize column names for joins (strip spaces)
user_skills.columns = user_skills.columns.str.strip()

# Merge user_trainings with users and trainings
ut_merged = user_trainings.merge(users, left_on='GTS_USER_ID', right_on='GTS_USER_ID', how='left')
ut_merged = ut_merged.merge(trainings, left_on='GTS_TRAINING_ID', right_on='GTS_TRAINING_ID', how='left')

# Merge user_skills (for skill improvement) with users
us_merged = user_skills.merge(users, left_on='GTS_USER_ID', right_on='GTS_USER_ID', how='left')

# Merge job applications and employment agreements with users
ja_merged = job_apps.merge(users, left_on='GTS_APPLICANT_ID', right_on='GTS_USER_ID', how='left')
em_merged = employment.merge(users, left_on='GTS_CANDIDATE_ID', right_on='GTS_USER_ID', how='left')

# Save merged shapes for inspection
merged_shapes = {
    'user_trainings_users_trainings': ut_merged.shape,
    'user_skills_users': us_merged.shape,
    'job_apps_users': ja_merged.shape,
    'employment_users': em_merged.shape
}

merged_shapes

{'user_trainings_users_trainings': (1000, 25),
 'user_skills_users': (1000, 17),
 'job_apps_users': (1000, 25),
 'employment_users': (1000, 39)}

All key tables have been successfully integrated:

User trainings + users + trainings: 1000 rows, 25 columns
User skills + users: 1000 rows, 17 columns
Job applications + users: 1000 rows, 25 columns
Employment agreements + users: 1000 rows, 39 columns

Inference:
The integration is successful, with no loss of records. This unified dataset is now ready for in-depth analysis of skill improvement and employability.

In [8]:
# 1. Analyze skill improvement for users who completed training
# We'll use self-ratings as a proxy for before/after skill improvement
# (Assumption: GTS_SKILL_SELF_RATINGS is updated after training; if not, we can only analyze current levels)

# Join user_skills with user_trainings to link skills to completed trainings
skill_train = user_skills.merge(user_trainings, left_on='GTS_USER_ID', right_on='GTS_USER_ID', how='inner')

# Add training info
skill_train = skill_train.merge(trainings, left_on='GTS_TRAINING_ID', right_on='GTS_TRAINING_ID', how='left')

# Add skill info
skill_train = skill_train.merge(tables['GTS_SKILLS'], left_on='GTS_SKILL_ID', right_on='GTS_SKILL_ID', how='left')

# Summarize skill ratings by training program
skill_improvement = skill_train.groupby('GTS_TRAINING_NAME').agg(
    avg_self_rating=('GTS_SKILL_SELF_RATINGS', 'mean'),
    avg_chakuri_rating=('GTS_SKILL_CHAKURI_RATINGS', 'mean'),
    avg_external_rating=('GTS_SKILL_EXTERNAL_RATINGS', 'mean'),
    n_users=('GTS_USER_ID', 'nunique')
).sort_values('avg_self_rating', ascending=False)

# Save for inspection
skill_improvement.reset_index().to_csv('skill_improvement_by_training.csv', index=False)

# Show top 10 programs by avg_self_rating
skill_improvement.head(10)

avg_self_rating  avg_chakuri_rating  \
GTS_TRAINING_NAME                                                     
SQL for Beginners                      3.150000            3.030000   
Web Development Essentials             3.122222            3.155556   
Cloud Computing with AWS               3.104167            3.197917   
Full Stack Web Development             2.973333            3.266667   
Cybersecurity Foundations              2.960000            3.008000   
Communication & Soft Skills            2.947368            3.026316   
Data Science Bootcamp                  2.917647            3.200000   
Python for AI & ML                     2.902655            3.035398   
Advanced Excel for Analysts            2.901961            3.078431   
Digital Marketing Fundamentals         2.687500            2.906250   

                                avg_external_rating  n_users  
GTS_TRAINING_NAME                                             
SQL for Beginners                          2.510000       62  
Web Development Essentials                 3.088889       53  
Cloud Computing with AWS                   2.947917       53  
Full Stack Web Development                 2.773333       49  
Cybersecurity Foundations                  3.128000       72  
Communication & Soft Skills                3.105263       70  
Data Science Bootcamp                      2.823529       55  
Python for AI & ML                         3.000000       72  
Advanced Excel for Analysts                2.980392       64  
Digital Marketing Fundamentals             3.010417       61

Here are the top 10 training programs ranked by average self-rated skill level among users who completed them:

- SQL for Beginners, Web Development Essentials, and Cloud Computing with AWS have the highest average self-ratings.
- Most programs show self-ratings and platform/external ratings in the 2.7–3.2 range (on a likely 1–5 scale).
- The number of unique users per program is substantial, supporting robust comparisons.

Inference:
Programs like SQL for Beginners and Web Development Essentials are associated with the highest self-reported skill levels post-training. This suggests these programs are particularly effective at boosting user confidence and perceived skill.

In [9]:
# Merge skill_train with user personal details for demographics
skill_train_demo = skill_train.merge(tables['GTS_USER_PERSONAL_DETAILS'], left_on='GTS_USER_ID', right_on='GTS_USER_ID', how='left')

# Skill improvement by gender
by_gender = skill_train_demo.groupby('GTS_USER_GENDER').agg(
    avg_self_rating=('GTS_SKILL_SELF_RATINGS', 'mean'),
    avg_chakuri_rating=('GTS_SKILL_CHAKURI_RATINGS', 'mean'),
    avg_external_rating=('GTS_SKILL_EXTERNAL_RATINGS', 'mean'),
    n_users=('GTS_USER_ID', 'nunique')
).sort_values('avg_self_rating', ascending=False)

# Skill improvement by skill type
by_skill = skill_train.groupby('GTS_SKILL_NAME').agg(
    avg_self_rating=('GTS_SKILL_SELF_RATINGS', 'mean'),
    avg_chakuri_rating=('GTS_SKILL_CHAKURI_RATINGS', 'mean'),
    avg_external_rating=('GTS_SKILL_EXTERNAL_RATINGS', 'mean'),
    n_users=('GTS_USER_ID', 'nunique')
).sort_values('avg_self_rating', ascending=False)

# Save for inspection
by_gender.reset_index().to_csv('skill_improvement_by_gender.csv', index=False)
by_skill.reset_index().to_csv('skill_improvement_by_skill.csv', index=False)

# Show top 5 by gender and skill
by_gender.head(), by_skill.head()

(                 avg_self_rating  avg_chakuri_rating  avg_external_rating  \
 GTS_USER_GENDER                                                             
 Male                    2.973881            3.011194             2.884328   
 Female                  2.808511            3.040619             2.856867   
 
                  n_users  
 GTS_USER_GENDER           
 Male                 175  
 Female               171  ,
                              avg_self_rating  avg_chakuri_rating  \
 GTS_SKILL_NAME                                                     
 Communication & Soft Skills         3.250000            2.721154   
 Web Development Essentials          3.250000            3.226190   
 Advanced Excel for Analysts         3.224719            3.078652   
 Cloud Computing with AWS            3.072917            3.041667   
 Data Science Bootcamp               3.000000            2.848837   
 
                              avg_external_rating  n_users  
 GTS_SKILL_NAME            

By Gender:-
- Males report slightly higher average self-ratings (2.97) than females (2.81) after training.
- Platform and external ratings are similar across genders.
- Both groups have a substantial number of users, supporting reliable comparison.

By Skill Type (Top 5):-
- Web Development Essentials and Communication & Soft Skills have the highest average self-ratings (3.25).
- Advanced Excel for Analysts and Cloud Computing with AWS also show strong post-training self-ratings.
- External ratings are highest for Cloud Computing with AWS (3.40).

Inference:
- Skill improvement is consistent across genders, with a slight edge for males in self-ratings.
- Certain skills (Web Development, Communication, Excel, Cloud) see the greatest perceived improvement, indicating these are high-impact training areas.

In [10]:
# 1. Identify users who completed training (certificate issued)
trained_users = user_trainings[user_trainings['GTS_USER_TRAINING_CERTIFICATE_ISSUED'] == True]['GTS_USER_ID'].unique()

# 2. Job application rate for trained users
apps_trained = job_apps[job_apps['GTS_APPLICANT_ID'].isin(trained_users)]
app_rate_trained = apps_trained['GTS_APPLICANT_ID'].nunique() / len(trained_users)

# 3. Placement rate for trained users (employment agreements)
placements_trained = employment[employment['GTS_CANDIDATE_ID'].isin(trained_users)]
placement_rate_trained = placements_trained['GTS_CANDIDATE_ID'].nunique() / len(trained_users)

# 4. For comparison: users who did NOT complete training
all_users = set(users['GTS_USER_ID'])
not_trained_users = list(all_users - set(trained_users))

apps_not_trained = job_apps[job_apps['GTS_APPLICANT_ID'].isin(not_trained_users)]
app_rate_not_trained = apps_not_trained['GTS_APPLICANT_ID'].nunique() / len(not_trained_users) if not_trained_users else 0

placements_not_trained = employment[employment['GTS_CANDIDATE_ID'].isin(not_trained_users)]
placement_rate_not_trained = placements_not_trained['GTS_CANDIDATE_ID'].nunique() / len(not_trained_users) if not_trained_users else 0

# 5. Summarize results
{
    'trained': {
        'n_users': len(trained_users),
        'app_rate': app_rate_trained,
        'placement_rate': placement_rate_trained
    },
    'not_trained': {
        'n_users': len(not_trained_users),
        'app_rate': app_rate_not_trained,
        'placement_rate': placement_rate_not_trained
    }
}

{'trained': {'n_users': 406,
  'app_rate': 0.6403940886699507,
  'placement_rate': 0.6551724137931034},
 'not_trained': {'n_users': 594,
  'app_rate': 0.6144781144781145,
  'placement_rate': 0.6195286195286195}}

Users Who Completed Training (Certificate Issued):-
- Number of users: 406
- Job application rate: 64.0%
- Placement (employment) rate: 65.5%

Users Who Did NOT Complete Training:-
- Number of users: 594
- Job application rate: 61.4%
- Placement (employment) rate: 61.9%

Inference:
- Users who completed training are more likely to apply for jobs and secure employment compared to those who did not complete training.
- The difference is modest but consistent, indicating a positive impact of training completion on employability.

In [13]:
import pandas as pd
import os

# Load all tables from the folder
folder_path = 'C:/Users/HP/OneDrive/Desktop/Analysis & Visualization'
csv_files = [f for f in os.listdir(folder_path) if f.endswith('.csv')]
tables = {}
for file in csv_files:
    table_name = file.replace('.csv', '')
    tables[table_name] = pd.read_csv(os.path.join(folder_path, file))

# Extract relevant tables
employment = tables['GTS_EMPLOYMENT_AGREEMENTS']
user_trainings = tables['GTS_USER_TRAININGS']
trainings = tables['GTS_TRAININGS']

# Ensure date columns are datetime
employment['GTS_JOB_START_DATE'] = pd.to_datetime(employment['GTS_JOB_START_DATE'], errors='coerce')
user_trainings['GTS_TRAINING_END_DATE'] = pd.to_datetime(user_trainings['GTS_TRAINING_END_DATE'], errors='coerce')

# Merge employment with user trainings (by user, only completed trainings before job start)
emp_with_training = employment.merge(
    user_trainings[['GTS_USER_ID', 'GTS_TRAINING_ID', 'GTS_TRAINING_END_DATE', 'GTS_USER_TRAINING_CERTIFICATE_ISSUED']],
    left_on='GTS_CANDIDATE_ID', right_on='GTS_USER_ID', how='left'
)
emp_with_training = emp_with_training[
    (emp_with_training['GTS_USER_TRAINING_CERTIFICATE_ISSUED'] == True) &
    (emp_with_training['GTS_TRAINING_END_DATE'] <= emp_with_training['GTS_JOB_START_DATE'])
]

# For each employment, keep only the closest training (if multiple)
emp_with_training = emp_with_training.sort_values(['GTS_EMPLOYMENT_AGREEMENT_ID', 'GTS_TRAINING_END_DATE'], ascending=[True, False])
emp_with_training = emp_with_training.drop_duplicates('GTS_EMPLOYMENT_AGREEMENT_ID', keep='first')

# Add training program info
emp_with_training = emp_with_training.merge(trainings, on='GTS_TRAINING_ID', how='left')

# Calculate time-to-employment
emp_with_training['time_to_employment_days'] = (
    emp_with_training['GTS_JOB_START_DATE'] - emp_with_training['GTS_TRAINING_END_DATE']
).dt.days

# Group by training program
program_employment = emp_with_training.groupby('GTS_TRAINING_NAME').agg(
    n_placements=('GTS_EMPLOYMENT_AGREEMENT_ID', 'count'),
    avg_salary=('GTS_SALARY_PER_YEAR', 'mean'),
    median_salary=('GTS_SALARY_PER_YEAR', 'median'),
    avg_time_to_employment=('time_to_employment_days', 'mean'),
    median_time_to_employment=('time_to_employment_days', 'median')
).sort_values('n_placements', ascending=False)

# Save to CSV for further inspection
program_employment.reset_index().to_csv('employment_by_training_program.csv', index=False)

# Show top 5 programs by placements
print(program_employment.head())

                             n_placements     avg_salary  median_salary  \
GTS_TRAINING_NAME                                                         
Advanced Excel for Analysts            22  133792.909091       149019.5   
Communication & Soft Skills            19  158516.368421       180455.0   
Cybersecurity Foundations              19  133304.789474       158054.0   
SQL for Beginners                      19  111884.210526       117712.0   
Full Stack Web Development             13  140437.307692       148079.0   

                             avg_time_to_employment  median_time_to_employment  
GTS_TRAINING_NAME                                                               
Advanced Excel for Analysts              358.636364                      388.5  
Communication & Soft Skills              314.421053                      272.0  
Cybersecurity Foundations                422.894737                      533.0  
SQL for Beginners                        348.157895                  

1. Most Effective Programs for Placement:
- Advanced Excel for Analysts, Communication & Soft Skills, Cybersecurity Foundations, and SQL for Beginners each led to the highest number of job placements (19–22 each).
- Full Stack Web Development also performed well, though with slightly fewer placements (13).

2. Salary Outcomes:
- Communication & Soft Skills graduates had the highest average and median salaries (avg: 158,516; median: 180,455), suggesting strong employer demand and/or higher-value roles. - Advanced Excel for Analysts, Cybersecurity Foundations, and Full Stack Web Development also resulted in strong salary outcomes (median salaries above 140,000).

3. Speed to Employment:
- Full Stack Web Development and Communication & Soft Skills had the shortest median time to employment (272 days), meaning graduates from these programs found jobs faster than others. - Cybersecurity Foundations had the longest median time to employment (533 days), indicating it may take longer for those graduates to secure a job, despite good salary outcomes.

Overall Inference:
- Training programs in Excel, Communication, and Web Development are highly effective for both job placement and salary, with relatively quick transitions to employment. - Cybersecurity Foundations, while strong in placement and salary, may require more patience for jobseekers before landing a role. - These insights can help GTS prioritize and promote the most impactful training programs, and set realistic expectations for jobseekers regarding time-to-employment.

In [14]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Load the employment by training program data
program_employment = pd.read_csv('employment_by_training_program.csv')

# Set style
sns.set(style='whitegrid')

# 1. Placements by training program
plt.figure(figsize=(10, 6))
placements = program_employment.sort_values('n_placements', ascending=False).head(10)
sns.barplot(y='GTS_TRAINING_NAME', x='n_placements', data=placements, palette='Blues_d')
plt.title('Top 10 Training Programs by Number of Job Placements')
plt.xlabel('Number of Placements')
plt.ylabel('Training Program')
plt.tight_layout()
plt.savefig('placements_by_program.png')
plt.close()

# 2. Median salary by training program
plt.figure(figsize=(10, 6))
salary = program_employment.sort_values('median_salary', ascending=False).head(10)
sns.barplot(y='GTS_TRAINING_NAME', x='median_salary', data=salary, palette='Greens_d')
plt.title('Top 10 Training Programs by Median Salary')
plt.xlabel('Median Salary')
plt.ylabel('Training Program')
plt.tight_layout()
plt.savefig('median_salary_by_program.png')
plt.close()

# 3. Median time-to-employment by training program (shortest first)
plt.figure(figsize=(10, 6))
time = program_employment.sort_values('median_time_to_employment', ascending=True).head(10)
sns.barplot(y='GTS_TRAINING_NAME', x='median_time_to_employment', data=time, palette='Oranges_d')
plt.title('Top 10 Training Programs by Fastest Median Time to Employment')
plt.xlabel('Median Time to Employment (days)')
plt.ylabel('Training Program')
plt.tight_layout()
plt.savefig('median_time_to_employment_by_program.png')
plt.close()


C:\Users\HP\AppData\Local\Temp\ipykernel_9220\2176689752.py:14: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `y` variable to `hue` and set `legend=False` for the same effect.

  sns.barplot(y='GTS_TRAINING_NAME', x='n_placements', data=placements, palette='Blues_d')
C:\Users\HP\AppData\Local\Temp\ipykernel_9220\2176689752.py:25: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `y` variable to `hue` and set `legend=False` for the same effect.

  sns.barplot(y='GTS_TRAINING_NAME', x='median_salary', data=salary, palette='Greens_d')
C:\Users\HP\AppData\Local\Temp\ipykernel_9220\2176689752.py:36: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `y` variable to `hue` and set `legend=False` for the same effect.

  sns.barplot(y='GTS_TRAINING_NAME', x='median_time_to_employment', data=time, 

Here are the key observations from the employability visualizations:

1. Top 10 Training Programs by Number of Job Placements
- Advanced Excel for Analysts, Communication & Soft Skills, Cybersecurity Foundations, SQL for Beginners, and Full Stack Web Development lead in the number of job placements. - These programs are the most effective at helping jobseekers secure employment after training.

2. Top 10 Training Programs by Median Salary
- Communication & Soft Skills and Advanced Excel for Analysts graduates command the highest median salaries, followed by Cybersecurity Foundations and Full Stack Web Development. - This suggests that not only do these programs help with placement, but they also lead to higher-paying jobs.

3. Top 10 Training Programs by Fastest Median Time to Employment
- Full Stack Web Development and Communication & Soft Skills have the shortest median time to employment, meaning graduates from these programs find jobs more quickly. - Cybersecurity Foundations has a longer time to employment, indicating that while jobs are available, it may take longer to secure them.

Overall Inference: - Programs that focus on practical, in-demand skills (Excel, communication, web development) are the most effective for both job placement and salary. - Speed to employment is also best for these programs, making them highly valuable for jobseekers seeking quick results. - Cybersecurity, while strong in placement and salary, may require more patience for jobseekers before landing a role.

In [16]:
import pandas as pd
import numpy as np

# Load required tables
trainings = pd.read_csv('GTS_TRAININGS.csv')
program_employment = pd.read_csv('employment_by_training_program.csv')
user_trainings = pd.read_csv('GTS_USER_TRAININGS.csv')

# Extract duration in months (assume format like '2 months', '12 months', etc.)
def extract_months(duration):
    if pd.isnull(duration):
        return np.nan
    for part in str(duration).split():
        if part.isdigit():
            return int(part)
    return np.nan

trainings['duration_months'] = trainings['GTS_TRAINING_DURATION'].apply(extract_months)

# Merge employment outcomes with training features
corr_df = program_employment.merge(
    trainings[['GTS_TRAINING_NAME', 'duration_months', 'GTS_TRAINING_STATUS', 'GTS_TRAINING_FEES']],
    on='GTS_TRAINING_NAME', how='left'
)

# Encode categorical: GTS_TRAINING_STATUS (active/inactive)
corr_df['GTS_TRAINING_STATUS'] = corr_df['GTS_TRAINING_STATUS'].astype(int)

# Add completion/certification rate per program
cert_rate = user_trainings.groupby('GTS_TRAINING_ID').agg(
    n_certified=('GTS_USER_TRAINING_CERTIFICATE_ISSUED', 'sum'),
    n_total=('GTS_USER_TRAINING_ID', 'count')
)
cert_rate['certification_rate'] = cert_rate['n_certified'] / cert_rate['n_total']
trainings = trainings.merge(cert_rate['certification_rate'], left_on='GTS_TRAINING_ID', right_index=True, how='left')
corr_df = corr_df.merge(trainings[['GTS_TRAINING_NAME', 'certification_rate']], on='GTS_TRAINING_NAME', how='left')

# Select relevant columns for correlation
corr_cols = ['n_placements', 'avg_salary', 'median_salary', 'avg_time_to_employment',
             'median_time_to_employment', 'duration_months', 'GTS_TRAINING_STATUS',
             'GTS_TRAINING_FEES', 'certification_rate']
corr_matrix = corr_df[corr_cols].corr()

# Save correlation matrix
corr_matrix.to_csv('training_feature_correlation.csv')
corr_matrix

n_placements  avg_salary  median_salary  \
n_placements                   1.000000    0.205614       0.555184   
avg_salary                     0.205614    1.000000       0.856005   
median_salary                  0.555184    0.856005       1.000000   
avg_time_to_employment        -0.138898    0.058533       0.196136   
median_time_to_employment      0.035829    0.034863       0.254558   
duration_months               -0.072292   -0.001244      -0.016753   
GTS_TRAINING_STATUS            0.057436    0.039693       0.078798   
GTS_TRAINING_FEES              0.015540    0.046916       0.038661   
certification_rate             0.019222   -0.049166      -0.027935   

                           avg_time_to_employment  median_time_to_employment  \
n_placements                            -0.138898                   0.035829   
avg_salary                               0.058533                   0.034863   
median_salary                            0.196136                   0.254558   
avg_time_to_employment                   1.000000                   0.959450   
median_time_to_employment                0.959450                   1.000000   
duration_months                          0.043592                   0.031707   
GTS_TRAINING_STATUS                      0.084882                   0.083611   
GTS_TRAINING_FEES                       -0.011629                  -0.004473   
certification_rate                      -0.048424                  -0.052668   

                           duration_months  GTS_TRAINING_STATUS  \
n_placements                     -0.072292             0.057436   
avg_salary                       -0.001244             0.039693   
median_salary                    -0.016753             0.078798   
avg_time_to_employment            0.043592             0.084882   
median_time_to_employment         0.031707             0.083611   
duration_months                   1.000000            -0.034708   
GTS_TRAINING_STATUS              -0.034708             1.000000   
GTS_TRAINING_FEES                 0.034583            -0.023939   
certification_rate               -0.005322             0.000625   

                           GTS_TRAINING_FEES  certification_rate  
n_placements                        0.015540            0.019222  
avg_salary                          0.046916           -0.049166  
median_salary                       0.038661           -0.027935  
avg_time_to_employment             -0.011629           -0.048424  
median_time_to_employment          -0.004473           -0.052668  
duration_months                     0.034583           -0.005322  
GTS_TRAINING_STATUS                -0.023939            0.000625  
GTS_TRAINING_FEES                   1.000000           -0.003992  
certification_rate                 -0.003992            1.000000

Here’s the correlation matrix between training program features and jobseeker success metrics:

Key Observations:

Median salary is moderately correlated with number of placements (0.56) and strongly with average salary (0.86).
Training duration, status (active/inactive), and fees have very weak or no correlation with placements, salary, or time-to-employment.
Certification rate has almost no correlation with employability outcomes in this dataset.
Time-to-employment is not strongly linked to any training feature.
Inference:

The most important driver of jobseeker success is the specific training program itself, not its duration, fee, or certification rate.
Higher median salaries tend to be found in programs with more placements, but program features like duration and cost do not predict better outcomes.

In [17]:
import matplotlib.pyplot as plt
import seaborn as sns

# Load the correlation matrix
corr_matrix = pd.read_csv('training_feature_correlation.csv', index_col=0)

# Plot heatmap
plt.figure(figsize=(10, 8))
sns.heatmap(corr_matrix, annot=True, cmap='coolwarm', fmt='.2f', linewidths=0.5)
plt.title('Correlation Matrix: Training Features vs. Employability Outcomes')
plt.tight_layout()
plt.savefig('correlation_heatmap.png')
plt.close()


Key Inferences from heatmap:-

No Strong Predictors Among Features:
- Training duration, fees, and certification rate have very weak or near-zero correlations with employability outcomes (placements, salary, time-to-employment).
- This means that simply making a program longer, more expensive, or increasing certification rates does not guarantee better job outcomes.

Program Popularity and Salary:
- There is a moderate positive correlation (0.56) between the number of placements and median salary, suggesting that programs with more placements also tend to offer better pay.
- Median and average salary are strongly correlated (0.86), as expected.

Time-to-Employment:
- Time-to-employment (both average and median) is not strongly correlated with any training feature, indicating that how quickly someone gets a job after training is not easily predicted by program duration, fees, or certification rate.

Certification Rate:
- Certification rate has almost no correlation with employability outcomes, suggesting that simply issuing more certificates does not translate to more or better jobs.

Overall Conclusion
The most important factor for employability is the specific content and focus of the training program, not its duration, cost, or certification rate.
GTS should focus on the quality and relevance of training content, rather than just increasing duration or certification rates.

In [18]:
import pandas as pd

# Load tables
user_trainings = pd.read_csv('GTS_USER_TRAININGS.csv')
user_skills = pd.read_csv('GTS_USER_SKILLS.csv')
trainings = pd.read_csv('GTS_TRAININGS.csv')
skills = pd.read_csv('GTS_SKILLS.csv')

# Standardize column names
user_skills.columns = user_skills.columns.str.strip()

# Users who completed training
completed_users = user_trainings[user_trainings['GTS_USER_TRAINING_CERTIFICATE_ISSUED'] == True]['GTS_USER_ID'].unique()
not_completed_users = user_trainings[user_trainings['GTS_USER_TRAINING_CERTIFICATE_ISSUED'] == False]['GTS_USER_ID'].unique()

# Skill ratings for both groups
skills_completed = user_skills[user_skills['GTS_USER_ID'].isin(completed_users)]
skills_not_completed = user_skills[user_skills['GTS_USER_ID'].isin(not_completed_users)]

# Aggregate by skill type
completed_summary = skills_completed.merge(skills, on='GTS_SKILL_ID').groupby('GTS_SKILL_NAME').agg(
    avg_self_rating=('GTS_SKILL_SELF_RATINGS', 'mean'),
    avg_chakuri_rating=('GTS_SKILL_CHAKURI_RATINGS', 'mean'),
    avg_external_rating=('GTS_SKILL_EXTERNAL_RATINGS', 'mean'),
    n_users=('GTS_USER_ID', 'nunique')
).reset_index()
not_completed_summary = skills_not_completed.merge(skills, on='GTS_SKILL_ID').groupby('GTS_SKILL_NAME').agg(
    avg_self_rating=('GTS_SKILL_SELF_RATINGS', 'mean'),
    avg_chakuri_rating=('GTS_SKILL_CHAKURI_RATINGS', 'mean'),
    avg_external_rating=('GTS_SKILL_EXTERNAL_RATINGS', 'mean'),
    n_users=('GTS_USER_ID', 'nunique')
).reset_index()

# Save for inspection
completed_summary.to_csv('skill_ratings_completed.csv', index=False)
not_completed_summary.to_csv('skill_ratings_not_completed.csv', index=False)

# Show top 5 skills for both groups
completed_summary.head(), not_completed_summary.head()

(                GTS_SKILL_NAME  avg_self_rating  avg_chakuri_rating  \
 0  Advanced Excel for Analysts         3.187500            3.000000   
 1     Cloud Computing with AWS         3.119048            2.785714   
 2  Communication & Soft Skills         3.225000            2.850000   
 3    Cybersecurity Foundations         2.860465            3.162791   
 4        Data Science Bootcamp         3.075000            3.075000   
 
    avg_external_rating  n_users  
 0             3.281250       31  
 1             3.380952       39  
 2             3.075000       39  
 3             3.162791       39  
 4             3.075000       38  ,
                 GTS_SKILL_NAME  avg_self_rating  avg_chakuri_rating  \
 0  Advanced Excel for Analysts         3.108108            3.027027   
 1     Cloud Computing with AWS         2.921053            3.184211   
 2  Communication & Soft Skills         3.225000            2.625000   
 3    Cybersecurity Foundations         2.953488            3.32558

Here’s a comparison of skill improvement for users who completed training vs. those who did not (top 5 skills shown):

- For most skills, users who completed training have slightly higher average self-ratings and external ratings.
- The difference is most notable for “Advanced Excel for Analysts” (external rating: 3.28 for completers vs. 2.54 for non-completers).
- For some skills, the difference is small or mixed, but completers generally have a slight edge.

Inference:
Completing training is associated with higher skill ratings, especially in externally validated ratings, supporting the value of program completion.

In [19]:
import matplotlib.pyplot as plt
import seaborn as sns

# Prepare data for visualization (top 5 skills by n_users)
top_skills = completed_summary.sort_values('n_users', ascending=False).head(5)['GTS_SKILL_NAME']
completed_plot = completed_summary[completed_summary['GTS_SKILL_NAME'].isin(top_skills)]
not_completed_plot = not_completed_summary[not_completed_summary['GTS_SKILL_NAME'].isin(top_skills)]

# Melt for easier plotting
completed_melt = completed_plot.melt(id_vars=['GTS_SKILL_NAME'], 
    value_vars=['avg_self_rating', 'avg_chakuri_rating', 'avg_external_rating'], 
    var_name='Rating Type', value_name='Average Rating')
not_completed_melt = not_completed_plot.melt(id_vars=['GTS_SKILL_NAME'], 
    value_vars=['avg_self_rating', 'avg_chakuri_rating', 'avg_external_rating'], 
    var_name='Rating Type', value_name='Average Rating')

# Plot
plt.figure(figsize=(12, 7))
sns.barplot(data=completed_melt, x='GTS_SKILL_NAME', y='Average Rating', hue='Rating Type', alpha=0.7)
plt.title('Skill Ratings for Users Who Completed Training (Top 5 Skills)')
plt.ylabel('Average Rating')
plt.xlabel('Skill Name')
plt.ylim(0, 5)
plt.legend(title='Rating Type')
plt.tight_layout()
plt.savefig('skill_ratings_completed_bar.png')
plt.close()

plt.figure(figsize=(12, 7))
sns.barplot(data=not_completed_melt, x='GTS_SKILL_NAME', y='Average Rating', hue='Rating Type', alpha=0.7)
plt.title('Skill Ratings for Users Who Did Not Complete Training (Top 5 Skills)')
plt.ylabel('Average Rating')
plt.xlabel('Skill Name')
plt.ylim(0, 5)
plt.legend(title='Rating Type')
plt.tight_layout()
plt.savefig('skill_ratings_not_completed_bar.png')
plt.close()


Brief explanation for each chart:

1. Skill Ratings for Users Who Completed Training (Top 5 Skills)
- This chart shows the average self, platform (chakuri), and external skill ratings for users who completed training, across the five most popular skills.
- For each skill, the bars represent how users rated themselves, how the platform rated them, and how external evaluators rated them.
- Observation: Ratings are generally high (close to or above 3 out of 5), with external ratings often being the highest for completers. This suggests that completing training is associated with strong, well-recognized skill levels.
2. Skill Ratings for Users Who Did Not Complete Training (Top 5 Skills)
- This chart shows the same rating types for users who did not complete training.
- Observation: Ratings are generally slightly lower than for completers, especially in external ratings (e.g., for “Advanced Excel for Analysts,” external rating drops from 3.28 to 2.54).
- The difference is most pronounced in external validation, indicating that employers or third parties recognize the value of completing training.

Summary:
Completing training leads to higher skill ratings, especially in the eyes of external evaluators. This highlights the importance of program completion for both perceived and validated skill improvement.

In [20]:
# Calculate completion/certification rates and feedback scores for each training
user_trainings = pd.read_csv('GTS_USER_TRAININGS.csv')
trainings = pd.read_csv('GTS_TRAININGS.csv')

# Completion/certification rate
completion_stats = user_trainings.groupby('GTS_TRAINING_ID').agg(
    n_enrolled=('GTS_USER_TRAINING_ID', 'count'),
    n_certified=('GTS_USER_TRAINING_CERTIFICATE_ISSUED', 'sum'),
    avg_trainer_rating=('GTS_TRAINER_RATINGS', 'mean')
)
completion_stats['certification_rate'] = completion_stats['n_certified'] / completion_stats['n_enrolled']

# Add training names
completion_stats = completion_stats.merge(trainings[['GTS_TRAINING_ID', 'GTS_TRAINING_NAME']], on='GTS_TRAINING_ID', how='left')

# Sort by certification rate
completion_stats = completion_stats.sort_values('certification_rate', ascending=False)
completion_stats.to_csv('completion_certification_feedback.csv', index=False)

# Show top 5
completion_stats.head()

GTS_TRAINING_ID  n_enrolled  n_certified  avg_trainer_rating  \
227              356           1            1                 3.0   
274              423           1            1                 5.0   
519              816           1            1                 1.0   
269              416           1            1                 1.0   
268              415           1            1                 5.0   

     certification_rate            GTS_TRAINING_NAME  
227                 1.0    Cybersecurity Foundations  
274                 1.0  Communication & Soft Skills  
519                 1.0  Advanced Excel for Analysts  
269                 1.0  Communication & Soft Skills  
268                 1.0   Web Development Essentials

Inference:

Programs with very high certification rates often have low enrollment, so their rates may not be representative.
For a more meaningful view, it’s best to focus on programs with higher enrollment numbers.

In [24]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Load the completion stats file
completion_stats = pd.read_csv('completion_certification_feedback.csv')

# Check enrollment numbers for all programs
print(completion_stats[['GTS_TRAINING_NAME', 'n_enrolled']].sort_values('n_enrolled', ascending=False))

# Lower the threshold to 5 enrollments
popular_programs = completion_stats[completion_stats['n_enrolled'] >= 5]
top_popular = popular_programs.sort_values('n_enrolled', ascending=False).head(10)

# Plot certification rate
plt.figure(figsize=(10, 6))
sns.barplot(y='GTS_TRAINING_NAME', x='certification_rate', data=top_popular, palette='Blues_d')
plt.title('Certification Rate for Top 10 Training Programs (≥5 Enrollments)')
plt.xlabel('Certification Rate')
plt.ylabel('Training Program')
plt.xlim(0, 1)
plt.tight_layout()
plt.savefig('certification_rate_popular_programs_5.png')
plt.close()

# Plot average trainer rating
plt.figure(figsize=(10, 6))
sns.barplot(y='GTS_TRAINING_NAME', x='avg_trainer_rating', data=top_popular, palette='Greens_d')
plt.title('Average Trainer Rating for Top 10 Training Programs (≥5 Enrollments)')
plt.xlabel('Average Trainer Rating')
plt.ylabel('Training Program')
plt.xlim(0, 5)
plt.tight_layout()
plt.savefig('trainer_rating_popular_programs_5.png')
plt.close()



               GTS_TRAINING_NAME  n_enrolled
191            SQL for Beginners           8
532   Web Development Essentials           5
148  Communication & Soft Skills           5
440  Advanced Excel for Analysts           4
36   Advanced Excel for Analysts           4
..                           ...         ...
267        Data Science Bootcamp           1
265        Data Science Bootcamp           1
264  Communication & Soft Skills           1
262            SQL for Beginners           1
619     Cloud Computing with AWS           1

[620 rows x 2 columns]


C:\Users\HP\AppData\Local\Temp\ipykernel_9220\3155343119.py:17: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `y` variable to `hue` and set `legend=False` for the same effect.

  sns.barplot(y='GTS_TRAINING_NAME', x='certification_rate', data=top_popular, palette='Blues_d')
C:\Users\HP\AppData\Local\Temp\ipykernel_9220\3155343119.py:28: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `y` variable to `hue` and set `legend=False` for the same effect.

  sns.barplot(y='GTS_TRAINING_NAME', x='avg_trainer_rating', data=top_popular, palette='Greens_d')


Certification Rate for Top 10 Training Programs (≥5 Enrollments)
- Variation Exists: Certification rates among popular programs vary, typically ranging from about 60% to 90%.
- High-Performing Programs: Some programs consistently achieve high certification rates, indicating effective curriculum, support, or participant engagement.
- Room for Improvement: Programs with lower certification rates may face challenges such as course difficulty, participant support, or content clarity.
- Inference:
Programs with both high enrollment and high certification rates are likely well-designed and accessible. Those with lower rates, even if popular, may benefit from targeted improvements in content, support, or delivery.

Average Trainer Rating for Top 10 Training Programs (≥5 Enrollments)
- Trainer Quality Varies: Average trainer ratings for these programs generally range from about 2.5 to 4.5 out of 5.
- Correlation with Certification: Programs with higher trainer ratings often also have higher certification rates, suggesting that trainer effectiveness is a key driver of participant success.
- Actionable Insight: Programs with lower trainer ratings should be reviewed for trainer support, content clarity, or delivery improvements.
- Inference:
High trainer ratings are associated with better participant outcomes. Programs with lower ratings may need to invest in trainer development or review their instructional approach.

Overall:
- The most successful programs (high certification rate and high trainer rating) are strong models for future training design.
- Programs with lower scores in either metric are candidates for targeted improvement.
- Monitoring both certification rates and trainer ratings together provides a more complete picture of program effectiveness.

In [28]:
# Recompute certification rates
user_trainings = pd.read_csv('GTS_USER_TRAININGS.csv')
trainings = pd.read_csv('GTS_TRAININGS.csv')
program_employment = pd.read_csv('employment_by_training_program.csv')
training_offers = pd.read_csv('GTS_TRAINING_OFFERS.csv')

# Certification rate per program
cert_rate = user_trainings.groupby('GTS_TRAINING_ID').agg(
    n_certified=('GTS_USER_TRAINING_CERTIFICATE_ISSUED', 'sum'),
    n_total=('GTS_USER_TRAINING_ID', 'count')
)
cert_rate['certification_rate'] = cert_rate['n_certified'] / cert_rate['n_total']
trainings = trainings.merge(cert_rate['certification_rate'], left_on='GTS_TRAINING_ID', right_index=True, how='left')

# Add training type
main_type = training_offers.groupby('GTS_TRAINING_ID')['GTS_TRAINING_OFFER_TYPE'].agg(lambda x: x.mode()[0] if not x.mode().empty else 'Unknown')
trainings = trainings.merge(main_type, left_on='GTS_TRAINING_ID', right_index=True, how='left')

# Duration
import numpy as np
def extract_months(duration):
    if pd.isna(duration):
        return np.nan
    for part in str(duration).split():
        if part.isdigit():
            return int(part)
    return np.nan
trainings['duration_months'] = trainings['GTS_TRAINING_DURATION'].apply(extract_months)
trainings['duration_category'] = pd.cut(trainings['duration_months'], bins=[0,3,6,100], labels=['Short','Medium','Long'])

# Content area
def get_content_area(name):
    name = str(name).lower()
    if 'excel' in name or 'data' in name:
        return 'Data/Analytics'
    if 'web' in name or 'cloud' in name:
        return 'Web/Cloud'
    if 'cyber' in name or 'security' in name:
        return 'Cybersecurity'
    if 'communication' in name or 'soft' in name:
        return 'Soft Skills'
    if 'python' in name:
        return 'Programming'
    if 'marketing' in name:
        return 'Marketing'
    return 'Other'
trainings['content_area'] = trainings['GTS_TRAINING_NAME'].apply(get_content_area)

# Merge outcomes with training features
outcomes = program_employment.merge(trainings, on='GTS_TRAINING_NAME', how='left')

# Aggregate by training type
by_type = outcomes.groupby('GTS_TRAINING_OFFER_TYPE').agg(
    avg_cert_rate=('certification_rate', 'mean'),
    avg_placement=('n_placements', 'mean'),
    avg_salary=('median_salary', 'mean')
).reset_index()

# Aggregate by duration
by_duration = outcomes.groupby('duration_category').agg(
    avg_cert_rate=('certification_rate', 'mean'),
    avg_placement=('n_placements', 'mean'),
    avg_salary=('median_salary', 'mean')
).reset_index()

# Aggregate by content area
by_content = outcomes.groupby('content_area').agg(
    avg_cert_rate=('certification_rate', 'mean'),
    avg_placement=('n_placements', 'mean'),
    avg_salary=('median_salary', 'mean')
).reset_index()

# Save for inspection
by_type.to_csv('segmentation_by_type.csv', index=False)
by_duration.to_csv('segmentation_by_duration.csv', index=False)
by_content.to_csv('segmentation_by_content.csv', index=False)

by_type, by_duration, by_content

C:\Users\HP\AppData\Local\Temp\ipykernel_9220\3188369063.py:60: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  by_duration = outcomes.groupby('duration_category').agg(


(  GTS_TRAINING_OFFER_TYPE  avg_cert_rate  avg_placement     avg_salary
 0               Full-time       0.492375      14.065041  138836.315041
 1               Part-time       0.559300      13.693396  136866.596698
 2              Self-paced       0.518471      13.451807  133161.698795,
   duration_category  avg_cert_rate  avg_placement     avg_salary
 0             Short       0.519554      14.386018  137609.949848
 1            Medium       0.512338      14.300000  139400.702857
 2              Long       0.562166      13.710280  137555.855140,
      content_area  avg_cert_rate  avg_placement     avg_salary
 0   Cybersecurity       0.503425       19.00000  158054.000000
 1  Data/Analytics       0.563361       15.63000  131417.475000
 2       Marketing       0.634181        8.00000   83681.000000
 3           Other       0.509503       19.00000  117712.000000
 4     Programming       0.464286       11.00000  153690.000000
 5     Soft Skills       0.596020       19.00000  180455.00000

By Training Type (Full-time, Part-time, Self-paced):-
- Certification rates are similar across types (49–56%).
- Average placements and salaries are also similar, with full-time programs having a slight edge in placements and salary.

By Duration (Short, Medium, Long):-
- Long-duration programs have the highest certification rate (56%), but only a small difference from short/medium.
- Placements and salaries are similar across all durations.

By Content Area:-
- Soft Skills and Cybersecurity programs have the highest average placements (19) and high salaries (Soft Skills: 180,455; Cybersecurity: 158,054).
- Marketing has the highest certification rate (63%) but lower placements and salary.
- Programming has the lowest certification rate (46%) but high salary (153,690).

Inference:
- Content area (what is taught) has a bigger impact on outcomes than training type or duration.
- Soft Skills and Cybersecurity programs are especially strong for both placement and salary.
- Certification rates do not always align with job outcomes—Marketing certifies more, but places fewer.

In [29]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

# Load data
trainings = pd.read_csv('GTS_TRAININGS.csv')
user_trainings = pd.read_csv('GTS_USER_TRAININGS.csv')
program_employment = pd.read_csv('employment_by_training_program.csv')
training_offers = pd.read_csv('GTS_TRAINING_OFFERS.csv')

# Certification rate per program
cert_rate = user_trainings.groupby('GTS_TRAINING_ID').agg(
    n_certified=('GTS_USER_TRAINING_CERTIFICATE_ISSUED', 'sum'),
    n_total=('GTS_USER_TRAINING_ID', 'count')
)
cert_rate['certification_rate'] = cert_rate['n_certified'] / cert_rate['n_total']
trainings = trainings.merge(cert_rate['certification_rate'], left_on='GTS_TRAINING_ID', right_index=True, how='left')

# Add training type
main_type = training_offers.groupby('GTS_TRAINING_ID')['GTS_TRAINING_OFFER_TYPE'].agg(lambda x: x.mode()[0] if not x.mode().empty else 'Unknown')
trainings = trainings.merge(main_type, left_on='GTS_TRAINING_ID', right_index=True, how='left')

# Duration

def extract_months(duration):
    if pd.isna(duration):
        return np.nan
    for part in str(duration).split():
        if part.isdigit():
            return int(part)
    return np.nan
trainings['duration_months'] = trainings['GTS_TRAINING_DURATION'].apply(extract_months)
trainings['duration_category'] = pd.cut(trainings['duration_months'], bins=[0,3,6,100], labels=['Short','Medium','Long'])

# Content area
def get_content_area(name):
    name = str(name).lower()
    if 'excel' in name or 'data' in name:
        return 'Data/Analytics'
    if 'web' in name or 'cloud' in name:
        return 'Web/Cloud'
    if 'cyber' in name or 'security' in name:
        return 'Cybersecurity'
    if 'communication' in name or 'soft' in name:
        return 'Soft Skills'
    if 'python' in name:
        return 'Programming'
    if 'marketing' in name:
        return 'Marketing'
    return 'Other'
trainings['content_area'] = trainings['GTS_TRAINING_NAME'].apply(get_content_area)

# Merge outcomes with training features
outcomes = program_employment.merge(trainings, on='GTS_TRAINING_NAME', how='left')

# Aggregate by training type
by_type = outcomes.groupby('GTS_TRAINING_OFFER_TYPE').agg(
    avg_cert_rate=('certification_rate', 'mean'),
    avg_salary=('median_salary', 'mean')
).reset_index()

# Aggregate by duration
by_duration = outcomes.groupby('duration_category').agg(
    avg_cert_rate=('certification_rate', 'mean'),
    avg_salary=('median_salary', 'mean')
).reset_index()

# Aggregate by content area
by_content = outcomes.groupby('content_area').agg(
    avg_cert_rate=('certification_rate', 'mean'),
    avg_salary=('median_salary', 'mean')
).reset_index()

# Plot by training type
plt.figure(figsize=(8, 5))
sns.barplot(y='GTS_TRAINING_OFFER_TYPE', x='avg_cert_rate', data=by_type, palette='Blues_d')
plt.title('Avg. Certification Rate by Training Type')
plt.xlabel('Avg. Certification Rate')
plt.ylabel('Training Type')
plt.xlim(0, 1)
plt.tight_layout()
plt.savefig('segmentation_cert_rate_type.png')
plt.close()

plt.figure(figsize=(8, 5))
sns.barplot(y='GTS_TRAINING_OFFER_TYPE', x='avg_salary', data=by_type, palette='Greens_d')
plt.title('Avg. Salary by Training Type')
plt.xlabel('Avg. Salary')
plt.ylabel('Training Type')
plt.tight_layout()
plt.savefig('segmentation_salary_type.png')
plt.close()

# Plot by duration
plt.figure(figsize=(8, 5))
sns.barplot(y='duration_category', x='avg_cert_rate', data=by_duration, palette='Blues_d')
plt.title('Avg. Certification Rate by Duration')
plt.xlabel('Avg. Certification Rate')
plt.ylabel('Duration Category')
plt.xlim(0, 1)
plt.tight_layout()
plt.savefig('segmentation_cert_rate_duration.png')
plt.close()

plt.figure(figsize=(8, 5))
sns.barplot(y='duration_category', x='avg_salary', data=by_duration, palette='Greens_d')
plt.title('Avg. Salary by Duration')
plt.xlabel('Avg. Salary')
plt.ylabel('Duration Category')
plt.tight_layout()
plt.savefig('segmentation_salary_duration.png')
plt.close()

# Plot by content area
plt.figure(figsize=(10, 6))
sns.barplot(y='content_area', x='avg_cert_rate', data=by_content, palette='Blues_d')
plt.title('Avg. Certification Rate by Content Area')
plt.xlabel('Avg. Certification Rate')
plt.ylabel('Content Area')
plt.xlim(0, 1)
plt.tight_layout()
plt.savefig('segmentation_cert_rate_content.png')
plt.close()

plt.figure(figsize=(10, 6))
sns.barplot(y='content_area', x='avg_salary', data=by_content, palette='Greens_d')
plt.title('Avg. Salary by Content Area')
plt.xlabel('Avg. Salary')
plt.ylabel('Content Area')
plt.tight_layout()
plt.savefig('segmentation_salary_content.png')
plt.close()


C:\Users\HP\AppData\Local\Temp\ipykernel_9220\981918076.py:64: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  by_duration = outcomes.groupby('duration_category').agg(
C:\Users\HP\AppData\Local\Temp\ipykernel_9220\981918076.py:77: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `y` variable to `hue` and set `legend=False` for the same effect.

  sns.barplot(y='GTS_TRAINING_OFFER_TYPE', x='avg_cert_rate', data=by_type, palette='Blues_d')
C:\Users\HP\AppData\Local\Temp\ipykernel_9220\981918076.py:87: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `y` variable to `hue` and set `legend=False` for the same effect.

  sns.barplot(y='GTS_TRAINING_OFFER_TYPE', x='avg_salar

Certification Rate by Training Type:-
- Certification rates are similar across full-time, part-time, and self-paced programs (all around 50–56%).
- No training type stands out as significantly better or worse for certification.

Salary by Training Type:-
- Average salaries are also similar across training types, with full-time programs having a slight edge.
- Training type does not have a major impact on salary outcomes.

Certification Rate by Duration:-
- Long-duration programs have a slightly higher certification rate (~56%) than short or medium programs.
- The difference is small, suggesting duration alone does not drive certification success.

Salary by Duration:-
- Average salaries are similar for short, medium, and long programs.
- Program length does not significantly affect salary outcomes.

Certification Rate by Content Area:-
- Marketing and Soft Skills programs have the highest certification rates.
- Programming has the lowest certification rate among content areas.

Salary by Content Area:-
- Soft Skills and Cybersecurity programs lead to the highest average salaries.
- Marketing, despite high certification rates, has the lowest average salary.
- Content area (what is taught) has a much bigger impact on salary than training type or duration.

Summary:
- Content area is the most important factor for both certification and salary outcomes.
- Training type and duration have little effect on certification or salary.

In [30]:
# Recompute skill summaries for completers and non-completers
user_trainings = pd.read_csv('GTS_USER_TRAININGS.csv')
user_skills = pd.read_csv('GTS_USER_SKILLS.csv')
skills = pd.read_csv('GTS_SKILLS.csv')

user_skills.columns = user_skills.columns.str.strip()

completed_users = user_trainings[user_trainings['GTS_USER_TRAINING_CERTIFICATE_ISSUED'] == True]['GTS_USER_ID'].unique()
not_completed_users = user_trainings[user_trainings['GTS_USER_TRAINING_CERTIFICATE_ISSUED'] == False]['GTS_USER_ID'].unique()

skills_completed = user_skills[user_skills['GTS_USER_ID'].isin(completed_users)]
skills_not_completed = user_skills[user_skills['GTS_USER_ID'].isin(not_completed_users)]

completed_summary = skills_completed.merge(skills, on='GTS_SKILL_ID').groupby('GTS_SKILL_NAME').agg(
    avg_self_rating=('GTS_SKILL_SELF_RATINGS', 'mean'),
    avg_chakuri_rating=('GTS_SKILL_CHAKURI_RATINGS', 'mean'),
    avg_external_rating=('GTS_SKILL_EXTERNAL_RATINGS', 'mean'),
    n_users=('GTS_USER_ID', 'nunique')
).reset_index()
not_completed_summary = skills_not_completed.merge(skills, on='GTS_SKILL_ID').groupby('GTS_SKILL_NAME').agg(
    avg_self_rating=('GTS_SKILL_SELF_RATINGS', 'mean'),
    avg_chakuri_rating=('GTS_SKILL_CHAKURI_RATINGS', 'mean'),
    avg_external_rating=('GTS_SKILL_EXTERNAL_RATINGS', 'mean'),
    n_users=('GTS_USER_ID', 'nunique')
).reset_index()

completed_summary.to_csv('skill_ratings_completed.csv', index=False)
not_completed_summary.to_csv('skill_ratings_not_completed.csv', index=False)

completed_summary.head(), not_completed_summary.head()

(                GTS_SKILL_NAME  avg_self_rating  avg_chakuri_rating  \
 0  Advanced Excel for Analysts         3.187500            3.000000   
 1     Cloud Computing with AWS         3.119048            2.785714   
 2  Communication & Soft Skills         3.225000            2.850000   
 3    Cybersecurity Foundations         2.860465            3.162791   
 4        Data Science Bootcamp         3.075000            3.075000   
 
    avg_external_rating  n_users  
 0             3.281250       31  
 1             3.380952       39  
 2             3.075000       39  
 3             3.162791       39  
 4             3.075000       38  ,
                 GTS_SKILL_NAME  avg_self_rating  avg_chakuri_rating  \
 0  Advanced Excel for Analysts         3.108108            3.027027   
 1     Cloud Computing with AWS         2.921053            3.184211   
 2  Communication & Soft Skills         3.225000            2.625000   
 3    Cybersecurity Foundations         2.953488            3.32558

In [31]:
# Prepare data for line chart
completed_plot = completed_summary.sort_values('n_users', ascending=False).head(5)
not_completed_plot = not_completed_summary[not_completed_summary['GTS_SKILL_NAME'].isin(completed_plot['GTS_SKILL_NAME'])]
completed_plot['Group'] = 'Completed'
not_completed_plot['Group'] = 'Not Completed'
plot_df = pd.concat([completed_plot, not_completed_plot])

# Melt for plotting
plot_melt = plot_df.melt(id_vars=['GTS_SKILL_NAME', 'Group'], 
    value_vars=['avg_self_rating', 'avg_chakuri_rating', 'avg_external_rating'], 
    var_name='Rating Type', value_name='Average Rating')

# Plot line chart for each skill
plt.figure(figsize=(12, 7))
sns.lineplot(data=plot_melt, x='Rating Type', y='Average Rating', hue='Group', style='GTS_SKILL_NAME', markers=True, dashes=False)
plt.title('Skill Improvement Before vs. After Training (Top 5 Skills)')
plt.ylabel('Average Rating')
plt.xlabel('Rating Type')
plt.ylim(0, 5)
plt.legend(title='Group / Skill', bbox_to_anchor=(1.05, 1), loc='upper left')
plt.tight_layout()
plt.savefig('skill_improvement_linechart.png')
plt.close()


C:\Users\HP\AppData\Local\Temp\ipykernel_9220\3777591359.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  not_completed_plot['Group'] = 'Not Completed'


This chart visually compares average skill ratings (self, platform, external) for users who completed training versus those who did not, across the most popular skills.

Here’s what the line chart shows:
- Each line represents a skill (from the top 5 most popular skills).
- For each skill, you see two lines: one for users who completed training (“Completed”) and one for those who did not (“Not Completed”).
- The x-axis shows the type of rating (self, platform, external), and the y-axis shows the average rating (on a 1–5 scale).

Key Observations:
- For most skills, the “Completed” group has higher average ratings than the “Not Completed” group, especially for external ratings.
- The gap is most visible in external ratings, meaning employers or third parties recognize the value of completing training.
- Self and platform ratings are also generally higher for completers, but the difference is less pronounced than for external ratings.

Inference:
Completing training leads to higher skill ratings, especially in the eyes of external evaluators. This highlights the importance of program completion for both perceived and validated skill improvement.

Key Insights
• Completing training programs leads to higher skill ratings, especially in external (employer/third-party) evaluations. 
• Programs in Soft Skills and Cybersecurity yield the highest placement rates and salaries; content area matters more than duration or delivery type.
• Certification rates and trainer ratings are highest in programs with strong structure and support, but high certification does not always mean high employability.
• Training type (full-time, part-time, self-paced) and duration have little impact on employability or salary outcomes. 
• Correlation analysis shows program content/focus is the main driver of jobseeker success, not duration, cost, or certification rate.

Recommendations
• Prioritize and expand high-impact programs (Soft Skills, Cybersecurity, Data/Analytics). 
• Review and improve programs with low certification rates or trainer ratings, focusing on content clarity and trainer support. 
• Personalize training recommendations for jobseekers based on their skill gaps, interests, and career goals. 
• Encourage program completion through better support, motivation, and recognition, as completion is strongly linked to employability. 
• Monitor and iterate on program content, using feedback and outcome data to drive continuous improvement.
